hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/StakingPeriodTimeHook.java

```diff
@@ -0,0 +1,104 @@
+
+    private final EndOfStakingPeriodUpdater stakingCalculator;
+    private Instant consensusTimeOfLastHandledTxn;
+
+    @Inject
+    public StakingPeriodTimeHook(@NonNull final EndOfStakingPeriodUpdater stakingPeriodCalculator) {
+        this.stakingCalculator = requireNonNull(stakingPeriodCalculator);
+    }
+
+    /**
+     * This time hook is responsible for performing necessary staking updates and distributing staking
+     * rewards. This should only be done during handling of the first transaction of each new staking
+     * period, which staking period usually starts at midnight UTC.
+     *
+     * <p>The only exception to this rule is when {@code consensusTimeOfLastHandledTxn} is null,
+     * <b>which should only happen on node startup.</b> The node should therefore run this process
+     * to catch up on updates and distributions when first coming online.
+     */
+    @Override
+    public void process(@NonNull Instant consensusTime, @NonNull final HandleContext context) {
+        if (consensusTimeOfLastHandledTxn == null
+                || consensusTime.getEpochSecond() > consensusTimeOfLastHandledTxn.getEpochSecond()
+                        && isNextStakingPeriod(consensusTime, consensusTimeOfLastHandledTxn, context)) {
+            // Handle the daily staking distributions and updates
+            try {
+                stakingCalculator.updateNodes(consensusTime, context);
+            } catch (final Exception e) {
+                logger.error("CATASTROPHIC failure updating end-of-day stakes", e);
+            }
+
+            // Advance the last consensus time to the given consensus time
+            consensusTimeOfLastHandledTxn = consensusTime;
+        }
+    }
+
+    @VisibleForTesting
+    void setLastConsensusTime(@Nullable final Instant lastConsensusTime) {
+        consensusTimeOfLastHandledTxn = lastConsensusTime;
+    }
+
+    @VisibleForTesting
+    static boolean isNextStakingPeriod(
+            @NonNull final Instant currentConsensusTime,
+            @NonNull final Instant previousConsensusTime,
+            @NonNull final HandleContext handleContext) {
+        final var stakingPeriod =
+                handleContext.configuration().getConfigData(StakingConfig.class).periodMins();
+        if (stakingPeriod == DEFAULT_STAKING_PERIOD_MINS) {
+            return isLaterUtcDay(currentConsensusTime, previousConsensusTime);
+        } else {
+            return getPeriod(currentConsensusTime, stakingPeriod * MINUTES_TO_MILLISECONDS)
+                    > getPeriod(previousConsensusTime, stakingPeriod * MINUTES_TO_MILLISECONDS);
+        }
+    }
+
+    private static boolean isLaterUtcDay(@NonNull final Instant now, @NonNull final Instant then) {
+        final var nowDay = LocalDate.ofInstant(now, UTC);
+        final var thenDay = LocalDate.ofInstant(then, UTC);
+        return nowDay.isAfter(thenDay) && nowDay.getDayOfYear() != thenDay.getDayOfYear();
```

Comment:
Do we need `nowDay.getDayOfYear() != thenDay.getDayOfYear();` ?

Reply:
I would think for `LocalDate` that `isAfter(...)` should be sufficient. Wiser minds than mine, should we drop the dayOfYear comparison?

Reply:
yeah probably.

Reply:
Done

Reply:
LocalDate.isAfter compares by epoch day, so yes, it is sufficient for a UTC midnight comparison.